From aa7ec3712edb25c1dd72b131d624830eb65382f8 Mon Sep 17 00:00:00 2001
From: WANG Xuerui <git@xen0n.name>
Date: Thu, 5 Dec 2024 14:41:56 +0800
Subject: [PATCH 2/5] Enable TLSDESC and DT_RELR for loongarch64 builds

While at it, also explicitly spell out the `la64v1.0` ISA baseline.

Signed-off-by: WANG Xuerui <git@xen0n.name>
---
 src/diagnostics/eng/native/configurecompiler.cmake | 9 +++++++++
 src/runtime/eng/native/configurecompiler.cmake     | 9 +++++++++
 2 files changed, 18 insertions(+)

diff --git a/src/diagnostics/eng/native/configurecompiler.cmake b/src/diagnostics/eng/native/configurecompiler.cmake
index 312312daa31..aa6803d4bb6 100644
--- a/src/diagnostics/eng/native/configurecompiler.cmake
+++ b/src/diagnostics/eng/native/configurecompiler.cmake
@@ -341,6 +341,15 @@ elseif (CLR_CMAKE_HOST_ARCH_ARM64)
 elseif (CLR_CMAKE_HOST_ARCH_LOONGARCH64)
   set(ARCH_HOST_NAME loongarch64)
   add_definitions(-DHOST_LOONGARCH64 -DHOST_64BIT)
+  # dotnet-unofficial-build: no easy way to set {C,CXX,LD}FLAGS for $CHOST
+  # *only*; these leak to $CBUILD and break crosscomponents' x64 build
+  # so we resort to putting them here in the sources
+  add_compile_options(-march=la64v1.0)
+  add_compile_options(-mtls-dialect=desc)
+  add_linker_flag(-Wl,-O1)
+  add_linker_flag(-Wl,--as-needed)
+  add_linker_flag(-Wl,-z,pack-relative-relocs)
+  add_linker_flag(-Wl,--hash-style=gnu)
 elseif (CLR_CMAKE_HOST_ARCH_RISCV64)
   set(ARCH_HOST_NAME riscv64)
   add_definitions(-DHOST_RISCV64 -DHOST_64BIT)
diff --git a/src/runtime/eng/native/configurecompiler.cmake b/src/runtime/eng/native/configurecompiler.cmake
index 901f601c333..dcc733d5582 100644
--- a/src/runtime/eng/native/configurecompiler.cmake
+++ b/src/runtime/eng/native/configurecompiler.cmake
@@ -353,6 +353,15 @@ elseif (CLR_CMAKE_HOST_ARCH_ARM64)
 elseif (CLR_CMAKE_HOST_ARCH_LOONGARCH64)
   set(ARCH_HOST_NAME loongarch64)
   add_definitions(-DHOST_LOONGARCH64 -DHOST_64BIT)
+  # dotnet-unofficial-build: no easy way to set {C,CXX,LD}FLAGS for $CHOST
+  # *only*; these leak to $CBUILD and break crosscomponents' x64 build
+  # so we resort to putting them here in the sources
+  add_compile_options(-march=la64v1.0)
+  add_compile_options(-mtls-dialect=desc)
+  add_linker_flag(-Wl,-O1)
+  add_linker_flag(-Wl,--as-needed)
+  add_linker_flag(-Wl,-z,pack-relative-relocs)
+  add_linker_flag(-Wl,--hash-style=gnu)
 elseif (CLR_CMAKE_HOST_ARCH_RISCV64)
   set(ARCH_HOST_NAME riscv64)
   add_definitions(-DHOST_RISCV64 -DHOST_64BIT)
-- 
2.52.0

